jerry-ext/args/args.c

```diff
@@ -0,0 +1,97 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "jerryscript-ext/args.h"
+#include "jerryscript.h"
+
+/**
+ * It validate the JS arguments and assign them to the native argument.
+ *
+ * @return jerry undefined: the validator passes,
+ *         jerry error: the validator fails.
+ */
+jerry_value_t
+jerryx_validate_and_assign_args (const jerry_value_t *js_arg_p, /**< points to the js argument */
+                                 const jerry_length_t js_arg_cnt, /**< the count of the js arguments */
+                                 const jerryx_arg_t *c_arg_p, /**< points to the jerryx_arg_t */
+                                 const jerry_length_t c_arg_cnt) /**< the count of the native argument */
+{
+  jerry_value_t ret = jerry_create_undefined ();
+
+  if (c_arg_cnt == 0)
+  {
+    return ret;
+  }
+
+  jerryx_arg_iterator_t iterator =
```

Comment:
Thanks for the explanation. But that iterator is still bugging me. I couldn't stop thinking a bit and came up with another variant. Let me know what you think, whether I missed something again. (It's going to be a long example, sorry in advance. I've adapted that 2D point transform, too. Some comments in the code.)

```c
// note that this struct is just 64 bits
// not bigger than a double!
// should not cause trouble to return it from functions
typedef struct {
  jerry_value_t result;
  jerry_length_t consumed;
} jerryx_arg_result_t;

inline jerryx_arg_result_t
jerryx_arg_result_ok (jerry_length_t consumed)
{
  return (jerryx_arg_result_t)
  {
    .result = jerry_create_undefined (),
    .consumed = consumed
  }
}

inline jerryx_arg_result_t
jerryx_args_result_error (jerry_error_t error_type, const jerry_char_t *message_p)
{
  return (jerryx_arg_result_t)
  {
    .result = jerry_create_error (error_type, message_p),
    .consumed = 0
  }
}

// this for complex use cases perhaps?
inline jerryx_arg_result_t
jerryx_args_result_error_consumed (jerry_error_t error_type, const jerry_char_t *message_p, jerry_length_t consumed)
{
  return (jerryx_arg_result_t)
  {
    .result = jerry_create_error (error_type, message_p),
    .consumed = consumed
  }
}

typedef jerryx_arg_result_t (*jerryx_arg_transform_func_t) (const jerry_value_t *js_arg_p, /**< array of available JS
                                                                                            *   args */
                                                            jerry_length_t js_arg_cnt, /**< number of available args */
                                                            const jerryx_arg_t *c_arg_p); /**< native arg */

inline jerry_value_t
jerryx_arg_helper_get (const jerry_value_t *js_arg_p, jerry_length_t js_arg_cnt, jerry_length_t index)
{
  return index < js_arg_cnt ? js_arg_p[index] : jerry_create_undefined ();
}

static jerryx_arg_result_t
jerryx_arg_transform_boolean_strict_impl (const jerry_value_t *js_arg_p,
                                          jerry_length_t js_arg_cnt,
                                          const jerryx_arg_t *c_arg_p)
{
  jerry_value_t js_arg = jerryx_arg_helper_get (js_arg_p, js_arg_cnt, 0);

  if (!jerry_value_is_boolean (js_arg))
  {
    return jerryx_args_result_error (JERRY_ERROR_TYPE,
                                     (jerry_char_t *) "It is not a boolean.");
  }

  *(bool *)(c_arg_p->dest) = jerry_get_boolean_value (js_arg);

  return jerryx_arg_result_ok (1);
} /* jerryx_arg_transform_boolean_strict_impl */

static jerryx_arg_result_t
jerryx_arg_transform_boolean_impl (const jerry_value_t *js_arg_p,
                                   jerry_length_t js_arg_cnt,
                                   const jerryx_arg_t *c_arg_p)
{
  jerry_value_t js_arg = jerryx_arg_helper_get (js_arg_p, js_arg_cnt, 0);

  *(bool *)(c_arg_p->dest) = jerry_value_to_boolean (js_arg);

  return jerryx_arg_result_ok (1);
} /* jerryx_arg_transform_boolean_impl */

jerryx_arg_result_t
jerryx_arg_transform_args (const jerry_value_t *js_arg_p,
                           jerry_length_t js_arg_cnt,
                           const jerryx_arg_t *c_arg_p,
                           jerry_length_t c_arg_cnt)
{
  jerryx_arg_result_t ret = jerryx_arg_result_ok (0);

  for (; c_arg_cnt != 0 && !jerry_value_has_error_flag (ret.result); c_arg_cnt--, c_arg_p++)
  {
    jerryx_arg_result_t arg_ret = c_arg_p->func (c_arg_p, js_arg_p, js_arg_cnt);

    // FIXME? needs jerry_release_value (ret.result)? (current version does not release either)
    ret.result = arg_ret.result;
    ret.consumed += arg_ret.consumed;

    jerry_length_t arg_consumed = arg_ret.consumed > js_arg_cnt ? js_arg_cnt : arg_ret.consumed;
    js_arg_p += arg_consumed;
    js_arg_cnt -= arg_consumed;
  }

  return ret;
}

jerryx_arg_result_t
my_custom_gpoint_transform (const jerry_value_t *js_arg_p,
                            jerry_length_t js_arg_cnt,
                            const jerryx_arg_t *c_arg_p)
{
  Gpoint *dest = (Gpoint *) c_arg_p->dest;

  jerry_value_t x = jerryx_arg_helper_get (js_arg_p, js_arg_cnt, 0);
  jerry_value_t to_number = jerry_value_to_number (x);
  dest->x = jerry_get_number_value (to_number);
  jerry_release_value (to_number);

  jerry_value_t y = jerryx_arg_helper_get (js_arg_p, js_arg_cnt, 1);
  to_number = jerry_value_to_number (y);
  dest->y = jerry_get_number_value (to_number);
  jerry_release_value (to_number);

  return jerryx_arg_result_ok (2);
} 
```